src/extensions/default/CodeFolding/foldhelpers/foldgutter.js

```diff
@@ -0,0 +1,308 @@
+        @param {!number} linesDiff a number to show how many lines where removed or added to the document
+    */
+    function updateFoldsCache(cm, from, linesDiff) {
+        var range;
+        var minFoldSize = prefs.getSetting("minFoldSize") || 2;
+        var foldedLines = Object.keys(cm._lineFolds).map(function (d) {
+            return +d;
+        });
+        if (linesDiff === 0) {
+            if (foldedLines.indexOf(from) > -1) {
+                var opts = cm.state.foldGutter.options || {};
+                var rf = opts.rangeFinder || CodeMirror.fold.auto;
+                range = rf(cm, CodeMirror.Pos(from));
+
+                if (range && range.to.line - range.from.line >= minFoldSize) {
+                    cm._lineFolds[from] = range;
+                } else {
+                    delete cm._lineFolds[from];
+                }
+            }
+        } else if (foldedLines.length) {
+            var newFolds = {};
+            foldedLines.forEach(function (line) {
+                range = cm._lineFolds[line];
+                if (line < from || linesDiff === 0) {
+                    newFolds[line] = range;
+                } else {
+                    range.from.line = range.from.line + linesDiff;
+                    range.to.line = range.to.line + linesDiff;
+                    newFolds[line + linesDiff] = range;
+
+                }
+            });
+            cm._lineFolds = newFolds;
+        }
+    }
+
+    /**
+        Triggered when the content of the document changes. When the entire content of the document
+        is changed - e.g., changes made from a different editor, the same lineFolds are kept only if
+        they are still valid in the context of the new document content.
+
+        @param {!CodeMirror} cm the CodeMirror instance for the active editor
+        @param {!{origin:string, from: {ch:number, line:number}, to: {ch: number, line: number},
+            removed: string[], text: string[]}} changeObj detailed information about the change that occurred in the document
+    */
+    function onChange(cm, changeObj) {
+        if (changeObj.origin === "setValue") {//text content has changed outside of brackets
+            var folds = cm.getValidFolds(cm._lineFolds);
+            cm._lineFolds = folds;
+            Object.keys(folds).forEach(function (line) {
+                cm.foldCode(+line);
+            });
+        } else {
+            var state = cm.state.foldGutter;
+            var lineChanges = changeObj.text.length - changeObj.removed.length;
+            //update the lineFolds cache
+            updateFoldsCache(cm, changeObj.from.line, lineChanges);
+            if (lineChanges !== 0) {
+                updateFoldInfo(cm, changeObj.from.line + lineChanges, changeObj.from.line + lineChanges + 1);
+            }
+            state.from = changeObj.from.line;
+            state.to = 0;
+            window.clearTimeout(state.changeUpdate);
```

Comment:
@MiguelCastillo Though I wonder if it'd be better to try to keep this as close as possible to the original CM code, with the goal of getting our few changes merged back upstream so we don't have to maintain a fork of it...